### ISSUE TITLE
Deficient affiliation is matched to organization

### ISSUE THREAD

**Issue by przemyslawjacewicz** [openaire/iis#1244]
This task is a subtask of parent task #1129 .

A deficient affiliation extracted by CERMINE for article with id `50|dedup_wf_001::6e9c8dd0ff2c33c49a207b43a3bdcae6` is matched to organization.   

The article has 3 affiliations extracted by CERMINE:
```json
[
  {
    "organization": "Department of Medical Oncology, The s econd a ffiliated h ospital of Zhejiang University s chool of Medicine",
    "countryname": null,
    "countrycode": null,
    "address": "h angzhou, Zhejiang Province, People's republic of china",
    "rawtext": "Department of Medical Oncology, The s econd a ffiliated h ospital of Zhejiang University s chool of Medicine, h angzhou, Zhejiang Province, People's republic of china"
  },
  {
    "organization": "cancer institute, Key laboratory of cancer Prevention and intervention, chinese n ational Ministry of education",
    "countryname": null,
    "countrycode": null,
    "address": "h angzhou, Zhejiang Province, People's republic of china",
    "rawtext": "cancer institute, Key laboratory of cancer Prevention and intervention, chinese n ational Ministry of education, h angzhou, Zhejiang Province, People's republic of china"
  },
  {
    "organization": "nlr",
    "countryname": null,
    "countrycode": null,
    "address": "Mlr, WBc, neutrophil, monocyte",
    "rawtext": "nlr, Mlr, WBc, neutrophil, monocyte"
  }
]
```

Affiliation number 3 is clearly deficient - `organization` field holds only an abbreviation and other fields hold data not related to any metadata concerning affiliation. Yet affiliation matching module matches this document with organizations having 'nlr' as short name, what can be seen in this table table showing a join of affiliation matching results with organizations:
```text
+--------------------+--------------------+-------------+--------------------+--------------------+---------+------------------+-----------+--------------------+
|          documentId|      organizationId|matchStrength|                  id|                name|shortName|       countryName|countryCode|          websiteUrl|
+--------------------+--------------------+-------------+--------------------+--------------------+---------+------------------+-----------+--------------------+
|50|dedup_wf_001::...|20|dedup_wf_001::...|        0.962|20|dedup_wf_001::...|STICHTING NATIONA...|      NLR|       Netherlands|         NL|   http://www.nlr.nl|
|50|dedup_wf_001::...|20|opendoar____::...|        0.962|20|opendoar____::...|Netherlands Aeros...|      NLR|       Netherlands|         NL|  http://www.nlr.nl/|
|50|dedup_wf_001::...|20|dedup_wf_001::...|        0.962|20|dedup_wf_001::...|National Library ...|      NLR|Russian Federation|         RU|http://www.nlr.ru...|
|50|dedup_wf_001::...|20|dedup_wf_001::...|   0.99623805|20|dedup_wf_001::...|    Cancer Institute|      WIA|             India|         IN|http://cancerinst...|
|50|dedup_wf_001::...|20|dedup_wf_001::...|        0.962|20|dedup_wf_001::...|Norsk Landbruksrå...|      NLR|            Norway|         NO|  http://www.nlr.no/|
|50|dedup_wf_001::...|20|dedup_wf_001::...|        0.962|20|dedup_wf_001::...|National Aerospac...|      NLR|       Netherlands|         NL|  http://www.nlr.nl/|
|50|dedup_wf_001::...|20|dedup_wf_001::...|        0.962|20|dedup_wf_001::...|North Little Rock...|      NLR|     United States|         US|  http://nlr.ar.gov/|
|50|dedup_wf_001::...|20|dedup_wf_001::...|        0.962|20|dedup_wf_001::...|Netherlands Lepro...|      NLR|       Netherlands|         NL|http://leprosyrel...|
+--------------------+--------------------+-------------+--------------------+--------------------+---------+------------------+-----------+--------------------+
```

We should check why a deficient affiliation such as the one above is matched to organizations sharing the same abbreviation. 
To do that we should check the processing of the article by affiliation matching module and find how module's matchers and joiners work together to result in matching. 

A possible outcome of this task can be an implementation change in affiliation matching module which could prevent the matching. This however must be done with caution not to remove any valid matches. It is also possible that this task will not result in any implementation changes because any change will result in high number of removed valid matches. The result can then be a decrease of match strength that is task #1245 .

Zeppelin note with data for this task is https://iis-cdh5-test-gw.ocean.icm.edu.pl/zeppelin/#/notebook/2G2QUYZRF .

**Comment by przemyslawjacewicz** [comment]
## Input analysis
I've taken the most recent PROD run and analyzed input in terms of affiliations similar to the deficient one. 

Affiliation matching input documents characteristics:
```scala
    val avroDataFrameSupport = new AvroDataFrameSupport(spark)
    val inputAvroAffPath = "/tmp/przemyslaw.jacewicz/GitHub1244/input/extracted_document_metadata"
    val aff = avroDataFrameSupport.read(inputAvroAffPath, ExtractedDocumentMetadata.SCHEMA$).cache()

    // total input document count
    aff.count() //10547240

    // total input document with affiliations count
    aff
      .where($"affiliations".isNotNull)
      .count() //6823832

    // total input affiliation count
    aff
      .select(explode($"affiliations"))
      .count() //17951304
```

Input affiliation is uniquely identified by document id and position within affiliation list, so for further analysis we need to attach position to affiliations:

```scala
    val array_zip_with_index: UserDefinedFunction = udf((elems: Seq[Row]) => Option(elems).map(_.zipWithIndex).orNull,
      ArrayType(StructType(Array(
        StructField("affiliation", SchemaConverters.toSqlType(Affiliation.SCHEMA$).dataType),
        StructField("index", DataTypes.IntegerType))))
    )

    val affWithPosition = aff
      .select($"id", array_zip_with_index($"affiliations").as("affiliations_with_position"))

    affWithPosition.printSchema()
/*
root
 |-- id: string (nullable = true)
 |-- affiliations_with_position: array (nullable = true)
 |    |-- element: struct (containsNull = true)
 |    |    |-- affiliation: struct (nullable = true)
 |    |    |    |-- organization: string (nullable = true)
 |    |    |    |-- countryName: string (nullable = true)
 |    |    |    |-- countryCode: string (nullable = true)
 |    |    |    |-- address: string (nullable = true)
 |    |    |    |-- rawText: string (nullable = false)
 |    |    |-- index: integer (nullable = true)
*/
``` 

The deficient affiliation is the characterized by a single word `organization` field together with missing `countryname` and `countrycode` fields. We can then count how many affiliations are similar to the deficient one:
```scala
    def isAffiliationWithOrganizationOnly(affiliation: Column) = affiliation("organization").isNotNull
      .and(affiliation("countryname").isNull).and(affiliation("countrycode").isNull)

    def isAffiliationOrganizationSingleWord(affiliation: Column) = not(affiliation("organization").contains(" "))

    val deficientAffWithPosition = affWithPosition
      .select($"id", explode($"affiliations_with_position").as("affiliation_with_position"))
      .where(isAffiliationWithOrganizationOnly($"affiliation_with_position.affiliation"))
      .where(isAffiliationOrganizationSingleWord($"affiliation_with_position.affiliation"))

    deficientAffWithPosition.count() //106712
```
We can now see that the deficient affiliations are only a small subset of the whole input: 106712 out of  17951304.

**Comment by przemyslawjacewicz** [comment]
## Intermediate results analysis
### Matchers
There are 5 versions of `AffOrgMatcher` matching affiliations to organizations. I've saved the intermediate results to check which matcher does the match with the deficient affiliation:
```scala
    val docOrgRelationMatcherResults = spark.read
      .parquet("/tmp/przemyslaw.jacewicz/GitHub1244/eu.dnetlib.iis.wf.affmatching.match.AffOrgMatcher@774f2734")
      .cache()

    docOrgRelationMatcherResults.count() //1211418
    docOrgRelationMatcherResults
      .where($"affiliation.documentId" === "50|dedup_wf_001::6e9c8dd0ff2c33c49a207b43a3bdcae6")
      .count() //0

    /*---------- P ----------*/

    val mainSectionHashBucketMatcher = spark.read
      .parquet("/tmp/przemyslaw.jacewicz/GitHub1244/eu.dnetlib.iis.wf.affmatching.match.AffOrgMatcher@269825b1")
      .cache()

    mainSectionHashBucketMatcherResults.count() //9508633
    mainSectionHashBucketMatcher
      .where($"affiliation.documentId" === "50|dedup_wf_001::6e9c8dd0ff2c33c49a207b43a3bdcae6")
      .count() //0

    /*---------- P ----------*/

    val alternativeNameMainSectionHashBucketMatcherResults = spark.read
      .parquet("/tmp/przemyslaw.jacewicz/GitHub1244/eu.dnetlib.iis.wf.affmatching.match.AffOrgMatcher@faaeb17")
      .cache()

    alternativeNameMainSectionHashBucketMatcherResults.count() //666131
    alternativeNameMainSectionHashBucketMatcherResults
      .where($"affiliation.documentId" === "50|dedup_wf_001::6e9c8dd0ff2c33c49a207b43a3bdcae6")
      .count() //0

    /*---------- P ----------*/

    val shortNameMainSectionHashBucketMatcherResults = spark.read
      .parquet("/tmp/przemyslaw.jacewicz/GitHub1244/eu.dnetlib.iis.wf.affmatching.match.AffOrgMatcher@746b69fb")
      .cache()

    shortNameMainSectionHashBucketMatcherResults.count() //443710
    shortNameMainSectionHashBucketMatcherResults
      .where($"affiliation.documentId" === "50|dedup_wf_001::6e9c8dd0ff2c33c49a207b43a3bdcae6")
      .count() //7
    shortNameMainSectionHashBucketMatcherResults
      .where($"affiliation.documentId" === "50|dedup_wf_001::6e9c8dd0ff2c33c49a207b43a3bdcae6")
      .show(truncate = false)
/*
+---------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------+-------------+
|affiliation                                                    |organization                                                                                                                               |matchStrength|
+---------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------+-------------+
|[50|dedup_wf_001::6e9c8dd0ff2c33c49a207b43a3bdcae6, 3, nlr, , ]|[20|dedup_wf_001::66b4d456022868a706fdaa9d511a8998, national library russia, nlr, russian federation, ru, nlr.ru/eng/, []]                 |0.962        |
|[50|dedup_wf_001::6e9c8dd0ff2c33c49a207b43a3bdcae6, 3, nlr, , ]|[20|dedup_wf_001::fa02c455756943e58e7fb08e5a7576d9, national aerospace laboratory, nlr, netherlands, nl, nlr.nl/, []]                      |0.962        |
|[50|dedup_wf_001::6e9c8dd0ff2c33c49a207b43a3bdcae6, 3, nlr, , ]|[20|dedup_wf_001::6e96f8f4cc96288f9772ba5f8a120631, stichting nationaal lucht en ruimtevaartlaboratorium, nlr, netherlands, nl, nlr.nl, []]|0.962        |
|[50|dedup_wf_001::6e9c8dd0ff2c33c49a207b43a3bdcae6, 3, nlr, , ]|[20|dedup_wf_001::351db7072f44f66a3d8ee47b88dac2ee, netherlands leprosy relief, nlr, netherlands, nl, leprosyrelief.org/, []]              |0.962        |
|[50|dedup_wf_001::6e9c8dd0ff2c33c49a207b43a3bdcae6, 3, nlr, , ]|[20|opendoar____::1841bc5964ea0085f2bb28cd6a043db7, netherlands aerospace centre, nlr, netherlands, nl, nlr.nl/, []]                       |0.962        |
|[50|dedup_wf_001::6e9c8dd0ff2c33c49a207b43a3bdcae6, 3, nlr, , ]|[20|dedup_wf_001::8541f5855b79a9a508bab2d16d3c9a5e, north little rock health department, nlr, united states, us, nlr.ar.gov/, []]          |0.962        |
|[50|dedup_wf_001::6e9c8dd0ff2c33c49a207b43a3bdcae6, 3, nlr, , ]|[20|dedup_wf_001::fd702fb85b83582332dfdc6c8cdeff18, norsk landbruksradgiving, nlr, norway, no, nlr.no/, []]                                |0.962        |
+---------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------+-------------+
*/

    /*---------- P ----------*/

    val firstWordsHashBucketMatcherResults = spark.read
      .parquet("/tmp/przemyslaw.jacewicz/GitHub1244/eu.dnetlib.iis.wf.affmatching.match.AffOrgMatcher@69083f6")
      .cache()

    firstWordsHashBucketMatcherResults.count() //3281092
    firstWordsHashBucketMatcherResults
      .where($"affiliation.documentId" === "50|dedup_wf_001::6e9c8dd0ff2c33c49a207b43a3bdcae6")
      .count() //1
    firstWordsHashBucketMatcherResults
      .where($"affiliation.documentId" === "50|dedup_wf_001::6e9c8dd0ff2c33c49a207b43a3bdcae6")
      .show(truncate = false)
/*
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------+-------------+
|affiliation                                                                                                                                                          |organization                                                                                                           |matchStrength|
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------+-------------+
|[50|dedup_wf_001::6e9c8dd0ff2c33c49a207b43a3bdcae6, 2, cancer institute, key laboratory cancer prevention and intervention, chinese n ational ministry education, , ]|[20|dedup_wf_001::9388b3d0055dfeb2c0339909c4be291a, cancer institute, wia, india, in, cancerinstitutewia.in/ciwia/, []]|0.99623805   |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------+-------------+
*/
``` 
So 7 matches are done by `shortNameMainSectionHashBucketMatcher` and 1 by `firstWordsHashBucketMatcher` and the deficient affiliation match is done done by `shortNameMainSectionHashBucketMatcher`.

### Deficient affiliation match
To verify if the deficient affiliation match is a generic case or an edge case I've checked how many affiliations similar to the deficient one are matched by each matcher:
```scala
    def joinMatcherResultsAndAffWithPosition(matcherResultsDF: DataFrame, affWithPositionDF: DataFrame): DataFrame = matcherResultsDF
      .join(affWithPositionDF, matcherResultsDF("affiliation.documentId") === affWithPositionDF("id") &&
        matcherResultsDF("affiliation.position") === affWithPositionDF("affiliation_with_position.index"))

    /*---------- P ----------*/

    val docOrgRelationMatcherResultsForDeficientAff = joinMatcherResultsAndAffWithPosition(
      docOrgRelationMatcherResults, deficientAffWithPosition)
    docOrgRelationMatcherResultsForDeficientAff.count() //1561

    /*---------- P ----------*/

    val mainSectionHashBucketMatcherResultsForDeficientAff = joinMatcherResultsAndAffWithPosition(
      mainSectionHashBucketMatcherResults, deficientAffWithPosition)
    mainSectionHashBucketMatcherResultsForDeficientAff.count() //22065

    /*---------- P ----------*/

    val alternativeNameMainSectionHashBucketMatcherResultsForDeficientAff = joinMatcherResultsAndAffWithPosition(
      alternativeNameMainSectionHashBucketMatcherResults, deficientAffWithPosition)
    alternativeNameMainSectionHashBucketMatcherResultsForDeficientAff.count() //2347

    /*---------- P ----------*/

    val shortNameMainSectionHashBucketMatcherResultsForDeficientAff = joinMatcherResultsAndAffWithPosition(
      shortNameMainSectionHashBucketMatcherResults, deficientAffWithPosition)
    shortNameMainSectionHashBucketMatcherResultsForDeficientAff.count() //7156

    /*---------- P ----------*/

    val firstWordsHashBucketMatcherResultsForDeficientAff = joinMatcherResultsAndAffWithPosition(
      firstWordsHashBucketMatcherResults, deficientAffWithPosition)
    firstWordsHashBucketMatcherResultsForDeficientAff.count() //19230
```
So there are 52359 matches between deficient affiliations and organizations with 7156 done by `shortNameMainSectionHashBucketMatcher`.  Affiliation and organization can be matched by different matchers so I've checked the count of unique matches:
```scala
    docOrgRelationMatcherResultsForDeficientAff.union(
      mainSectionHashBucketMatcherResultsForDeficientAff).union(
      alternativeNameMainSectionHashBucketMatcherResultsForDeficientAff).union(
      shortNameMainSectionHashBucketMatcherResultsForDeficientAff).union(
      firstWordsHashBucketMatcherResultsForDeficientAff)
      .groupBy($"affiliation.documentId", $"organization.id")
      .agg(count(lit("1")))
      .count() //36553
```
There are only 36553 unique matches of deficient affiliations propagated to output. Digging further I've also checked how many deficient affiliations matched by `shortNameMainSectionHashBucketMatcher` are also matched by other matchers:
```scala
    def joinMatherResults(matcherResultsDF1: DataFrame, matcherResultsDF2: DataFrame): DataFrame = matcherResultsDF1.as("left")
      .join(matcherResultsDF2.as("right"),
        matcherResultsDF1("affiliation.documentId") === matcherResultsDF2("affiliation.documentId") &&
          matcherResultsDF1("affiliation.position") === matcherResultsDF2("affiliation.position") &&
          matcherResultsDF1("organization.id") === matcherResultsDF2("organization.id"))

    joinMatherResults(shortNameMainSectionHashBucketMatcherResultsForDeficientAff, docOrgRelationMatcherResults).count() //54
    joinMatherResults(shortNameMainSectionHashBucketMatcherResultsForDeficientAff, mainSectionHashBucketMatcherResults).count() //1236
    joinMatherResults(shortNameMainSectionHashBucketMatcherResultsForDeficientAff, alternativeNameMainSectionHashBucketMatcherResults).count() //13
    joinMatherResults(shortNameMainSectionHashBucketMatcherResultsForDeficientAff, firstWordsHashBucketMatcherResults).count() //1230

    joinMatherResults(shortNameMainSectionHashBucketMatcherResultsForDeficientAff, docOrgRelationMatcherResults)
      .union(joinMatherResults(shortNameMainSectionHashBucketMatcherResultsForDeficientAff, mainSectionHashBucketMatcherResults))
      .union(joinMatherResults(shortNameMainSectionHashBucketMatcherResultsForDeficientAff, alternativeNameMainSectionHashBucketMatcherResults))
      .union(joinMatherResults(shortNameMainSectionHashBucketMatcherResultsForDeficientAff, firstWordsHashBucketMatcherResults))
      .select($"left.affiliation.documentId", $"left.organization.id", $"right.affiliation.documentId", $"right.organization.id")
      .count() //1169
```
So there are only 7156 matches between deficient affiliations and organizations done by `shortNameMainSectionHashBucketMatcher` of which 1169 are also done by other matchers. This means that 5987 matches are uniqe to `shortNameMainSectionHashBucketMatcher`.

Summarizing this analysis: there are 106712 affiliations similar to the deficient one in input, out of those affiliations 36553 are matched to organizations with 7156 matched by `shortNameMainSectionHashBucketMatcher` which is the matcher producing `wrong` matches. 5987 matches are unique to this matcher so removing any matches from this matcher should not have a significant impact on affiliation matching outcome.

**Comment by przemyslawjacewicz** [comment]
## The fix

A simple fix can be applied to the matching procedure
1. return empty hash from `MainSectionBucketHasher` when input organization name is one-word, that is does not contain a space character
1. filter out empty hashes of affiliations and organizations in `AffOrgHashBucketJoiner`

These two changes are relatively simple and should change only the `mainSectionHashBucketMatcher` and `shortNameMainSectionHashBucketMatcher` because both matchers use `MainSectionBucketHasher`.

## Comparison after applying the fix
The following tables contains the comparison of numbers characterizing matching procedure with more details for matchers affected by the fix:

| `docOrgRelationMatcher`                   | base    | fix     |
|-------------------------------------------|---------|---------|
| match count                               | 1211418 | 1211418 |
| deficient affiliations match count | 8920    | 8920    |

| `mainSectionHashBucketMatcher`            | base    | fix     |
|-------------------------------------------|---------|---------|
| match count                               | 9508633 | 9374534 |
| deficient affiliations match count | 37244   | 0       |
| distinct deficient affiliations match count        | 35849  | 0      |
| matcher unique distinct deficient affiliations match count | 35358  | 0      |


| `alternativeNameMainSectionHashBucketMatcher` | base   | fix    |
|-----------------------------------------------|--------|--------|
| match count                                   | 666131 | 666130 |
| deficient affiliations match count     | 1      | 0      |


| `shortNameMainSectionHashBucketMatcher`                   | base   | fix    |
|-----------------------------------------------------------|--------|--------|
| match count                                               | 443710 | 331671 |
| deficient affiliations match count                 | 93470  | 0      |
| distinct deficient affiliations match count        | 88476  | 0      |
| matcher unique distinct deficient affiliations match count | 75027  | 0      |


| `firstWordsHashBucketMatcher`             | base    | fix     |
|-------------------------------------------|---------|---------|
| match count                               | 3281092 | 3281092 |
| deficient affiliations match count | 123455  | 123455  |


| all matchers                                       | base   | fix    |
|----------------------------------------------------|--------|--------|
| deficient affiliations match count          | 263090 | 132375 |
| unique deficient affiliations match count | 203570 | 128032 |

| output | base     | fix     |
|--------|----------|---------|
| count  | 10081164 | 9933051 |

The number of output records drops by 148113 (c. 1.49%).

In summary, after applying the fix the number of (unique) deficient affiliations matched `mainSectionHashBucketMatcher` and `shortNameMainSectionHashBucketMatcher` to organization drops to 0. This means that the fix properly removes the matches between deficient affiliations and organizations. But the number of removed records is 148113 as opposed to 110385(=75027+35358) meaning that we have also removed 37728 other records. These additional removed records must be due to filtering out empty hashes of affiliations and organizations. This however is probably not a bad thing because matches based on empty hashes are probably invalid. Moreover 37728 records out of 10M is 0,37728%, so a small fraction.

**Comment by przemyslawjacewicz** [comment]
We came to a conclusion that the condition for single-word organization in affiliation is too strict and by applying it we may remove many valid matches. I've took another attempt at narrowing down the deficient affiliations, adding some conditions that I've seen in real data. So as a first approximation we can say that an affiliation is deficient if the following conditions hold 
* organization name is a single word
* country name and country code are missing
* affiliation is first or last in the affiliation list

Taking this dataset we can further narrow deficient affiliations down by adding a condition that not all affiliations should be deficient - this condition takes into account valid affiliations that are all abbreviations. For refrence I'm adding the code implementing the conditions
```scala
      val inputAvroAffPath = "/tmp/przemyslaw.jacewicz/GitHub1129/GitHub1244/input/extracted_document_metadata"
      val aff = avroDataFrameSupport.read(inputAvroAffPath, ExtractedDocumentMetadata.SCHEMA$).cache()

    // position starts at 1, must provide result schema because of Rows in output
    def array_zip_with_position(colName: String, positionColName: String = "position"): UserDefinedFunction =
      udf((elems: Seq[Row]) => Option(elems).map(_.zipWithIndex.map(x => x.copy(_2 = x._2 + 1))).orNull,
        ArrayType(StructType(Array(
          StructField(colName, SchemaConverters.toSqlType(Affiliation.SCHEMA$).dataType),
          StructField(positionColName, DataTypes.IntegerType))))
      )

    val affWithPosition = aff
      .select($"id", explode(array_zip_with_position("affiliation")($"affiliations")).as("affiliation_with_position"))

    def string_contains(chars: String*): UserDefinedFunction = udf((str: String) =>
      Option(str).exists(x => chars.exists(c => x.contains(c))), DataTypes.BooleanType
    )

    val x = aff
      .withColumn("x", explode(array_zip_with_position("affiliation")($"affiliations")))
      .withColumn("is_organization_deficient", not(string_contains(" ", "-")($"x.affiliation.organization")))
      .withColumn("is_affiliation_first", $"x.position" === 1)
      .withColumn("is_affiliation_last", $"x.position" === size(aff("affiliations")))
      .select(
        $"id",
        $"x.affiliation",
        $"x.position",
        $"is_organization_deficient",
        $"is_affiliation_first",
        $"is_affiliation_last",
        ($"is_organization_deficient" && ($"is_affiliation_first" || $"is_affiliation_last")).as("is_affiliation_deficient")
      )

  class BoolAnd extends UserDefinedAggregateFunction {
    override def inputSchema: StructType = StructType(Array(StructField("value", DataTypes.BooleanType)))

    override def bufferSchema: StructType = StructType(Array(StructField("buffer", DataTypes.BooleanType)))

    override def dataType: DataType = DataTypes.BooleanType

    override def deterministic: Boolean = true

    // true is a neutral element of AND
    override def initialize(buffer: MutableAggregationBuffer): Unit = {
      buffer(0) = true
    }

    override def update(buffer: MutableAggregationBuffer, input: Row): Unit = {
      buffer(0) = buffer.getAs[Boolean](0) && input.getAs[Boolean](0)
    }

    override def merge(buffer1: MutableAggregationBuffer, buffer2: Row): Unit = {
      buffer1(0) = buffer1.getAs[Boolean](0) && buffer2.getAs[Boolean](0)
    }

    override def evaluate(buffer: Row): Any = buffer.getAs[Boolean](0)
  }

    val bool_and = new BoolAnd

    val groupedByAllDeficient = x
      .select($"id", $"is_affiliation_deficient")
      .groupBy($"id").agg(bool_and($"is_affiliation_deficient").as("are_all_affiliations_deficient"))

    val xx = x
      .join(groupedByAllDeficient, Seq("id"))

    val z = xx
      .where($"affiliation.organization".isNotNull && $"affiliation.countryName".isNull && $"affiliation.countryCode".isNull)
      .where($"is_affiliation_deficient" && not($"are_all_affiliations_deficient"))

    z.count() //46882
```
So we can narrow down deficient affiliations to 46882. Unfortunately manual sampling of the resulting dataset showed that it still contains valid affiliations. 

In conclusion : we are dropping the idea of changes in affiliation matching procedure or addition of any filtering of deficient affiliations. To actually prevent the matching the best way is to fix affiliation extraction to not extract deficient affiliations at all. This however is quite complicated and is moved to distant future.
